api/v1/clusterwidenetworkpolicy_types.go

```diff
@@ -109,8 +116,86 @@ type EgressRule struct {
 	// empty or missing, this rule matches all destinations (traffic not restricted by
 	// destination). If this field is present and contains at least one item, this rule
 	// allows traffic only if the traffic matches at least one item in the to list.
+	// To rules can't contain ToFQDNs rules.
 	// +optional
 	To []networking.IPBlock `json:"to,omitempty"`
+
+	// List of FQDNs(fully qualified domain name) for outgoing traffic of a cluster for this rule.
+	// Items in this list are combined using a logical OR operation. This field is used as
+	// whitelist of DNS names. If none specified, rule will not be applied.
+	// ToFQDNs rules can't contain To rules.
+	// +optional
+	ToFQDNs []FQDNSelector `json:"toFQDNs,omitempty"`
+}
+
+// FQDNSelector describes rules for matching DNS names.
+type FQDNSelector struct {
+	// MatchName matches FQDN.
+	// +kubebuilder:validation:Pattern=`^([-a-zA-Z0-9_]+[.]?)+$`
+	MatchName string `json:"matchName,omitempty"`
+
+	// MatchPattern allows using "*" to match DNS names.
+	// "*" matches 0 or more valid characters.
+	// +kubebuilder:validation:Pattern=`^([-a-zA-Z0-9_*]+[.]?)+$`
+	MatchPattern string `json:"matchPattern,omitempty"`
+
+	// Sets stores nftables sets used for rule.
+	// Stored in Spec so that reconciler would receive always up-to-date data.
+	// Client caches Status, so additional actions would be required to guarantee that up-to-date data is used.
+	// +optional
+	Sets []IPSet `json:"sets,omitempty"`
+}
+
+// IPSet stores set name association to IP addresses
+type IPSet struct {
+	FQDN           string      `json:"fqdn,omitempty"`
+	SetName        string      `json:"setName,omitempty"`
+	IPs            []string    `json:"ips,omitempty"`
+	ExpirationTime metav1.Time `json:"expirationTime,omitempty"`
+	Version        IPVersion   `json:"version,omitempty"`
+}
+
+func (l *ClusterwideNetworkPolicyList) GetFQDNs() []FQDNSelector {
+	s := []FQDNSelector{}
+	for _, i := range l.Items {
+		for _, e := range i.Spec.Egress {
+			s = append(s, e.ToFQDNs...)
+		}
+	}
+
+	return s
+}
+
+func (s FQDNSelector) GetName() string {
+	if s.MatchName != "" {
+		return s.MatchName
+	}
+
+	return s.MatchPattern
+}
+
+func (s FQDNSelector) GetMatchName() string {
+	return dnsgo.Fqdn(s.MatchName)
+}
+
+// GetRegex converts a MatchPattern into a regexp string
+func (s FQDNSelector) GetRegex() string {
```

Comment:
Here is example that doesn't work well with `filepath.Match`: https://go.dev/play/p/GqUv60XjH2Q. Specifically, if pattern `example.*` is given then it thinks that `example.bar.com` is valid match. So if user would provide 2 patterns: `example.*` and `*.bar.com` -- they would conflict. While it could sound like an unlikely case i would still advocate for avoiding using `filepath.Match` as more such problematic cases could be uncovered.

Also, this regex implementation was inspired by Cilium project, i think they used similar regexes in their DNS-based policies. So they should be somewhat tested, i hope)

Reply:
I will add a few edge test cases.